File: backend/src/main/java/com/xxxx/systemvotting/modules/poll/service/CategoryService.java
package com.xxxx.systemvotting.modules.poll.service;

import com.xxxx.systemvotting.modules.poll.dto.CategoryDTO;

import java.util.List;

public interface CategoryService {

    /**
     * Returns all categories ordered by sortOrder ascending.
     */
    List<CategoryDTO> getAllCategories();

    /**
     * Finds a category by its slug.
     */
    CategoryDTO getCategoryBySlug(String slug);

    /**
     * Creates a new category.
     */
    CategoryDTO createCategory(CategoryDTO categoryDTO);

    /**
     * Updates an existing category.
     */
    CategoryDTO updateCategory(Long id, CategoryDTO categoryDTO);

    /**
     * Deletes a category and nullifies its reference in associated polls.
     */
    void deleteCategory(Long id);
}

File: frontend/src/services/category.service.ts
import api from './api';

export interface Category {
    id: number;
    name: string;
    slug: string;
    icon: string;  // emoji, e.g. "🔥", "💻"
    sortOrder: number;
    pollCount?: number;
}

export const categoryService = {
    getAllCategories: async (): Promise<Category[]> => {
        const response = await api.get('/categories');
        return response.data.data;
    },

    getCategoryBySlug: async (slug: string): Promise<Category> => {
        const response = await api.get(`/categories/${slug}`);
        return response.data.data;
    },

    createCategory: async (category: Partial<Category>): Promise<Category> => {
        const response = await api.post('/categories', category);
        return response.data.data;
    },

    updateCategory: async (id: number, category: Partial<Category>): Promise<Category> => {
        const response = await api.put(`/categories/${id}`, category);
        return response.data.data;
    },

    deleteCategory: async (id: number): Promise<void> => {
        await api.delete(`/categories/${id}`);
    },
};
